# Day 1 Exercises
## Data Types

#### 1. What types are involved in the following expressions? What data type will each expression evaluate to?


In [ ]:
8 * 2.5

an integer is involved in the expression which is 8;
float number is involved in the expression which is 2.5;
8 and 2.5 are considered operands; * is considered an operation which multiplies and evaluates the integers

In [ ]:
9 / 2

both 9 and 2 are considered integers which are then operands in the expression; / is considered an operation which divides and evaluates the integers

In [ ]:
1.5 * 2 >= 7 - 3

1.5 is considered a float number;
2,7, and 3 are considered intergers
"*" is an operator which will mulitply 1.5 and 2
"-" is an operator which will subtract 7 and 3
">=" is an operator that will evaluate whether the product of 1.5 and 2 is greater than or equal to the difference of 7 and 3

#### 2. In which order will the expressions be evaluated?

In [ ]:
6 * 3 + 7 * 4

first 6 and 3 would be mulitplied, then 7 and 4 would also be multiplied then the products from the previous expressions will be added together to have a total

In [ ]:
5 - 2 * 3 ** 4

3 would be evaluated to the power of 4, then the answer from the previous expression would be multiplied by 2, then the product from that expression would be subtracted by 5 to find the difference

In [ ]:
(5 - 2) * 3 ** 4

the expression 5-2 would be evaluated first, then 3 would be raised to the power of 4, then the difference from the first expression would be subtracted by the answer from the raised integer

In [ ]:
5 + 2 >= 3 * 4

3 would be multiplied by 4, then 5 would be added by 2, then the expression would evaulate if the sum from the first expression would be greater than or equal to the product of the second expression

## Comments and Errors

#### 1. Which of these expressions results in an error? Why does each error occur?

In [1]:
((((5 * 4 ** 7))))

81920

4 is raised to the power of 7 then is multiplied by 5; answer is 81920; does not result in an error

In [2]:
x = 84
x *= 0.5 / 7
print(x)

SyntaxError: 'literal' is an illegal expression for augmented assignment (868797654.py, line 1)

there is an error in the expression; number 84 cannot be a variable therefore the number has to be assigned to a variable like x

In [3]:
(-(-(-(-5 * (4 + 3)))))

35

there is no error; 4 would be added by 3 then multiplied by negative 5, then the entire expression would be multiplied by negative 1 to the power of 4

In [27]:
5 * 3 = weight

SyntaxError: cannot assign to expression here. Maybe you meant '==' instead of '='? (3700621706.py, line 1)

you cannot assign a variable beginning with a number it would result in an error; you would have to swap the order or put "==" instead of "="

In [5]:
73 / -----------5

-14.6

does not result in an error would divide 73 by negative 5

#### 2. Write a block of code to check if the value is a string. If it is, print out a message saying it is valid. Otherwise, raise an exception with a message saying the value must be a string.

In [6]:
value = 6

if type(value) == str:
    print("Valid")
else:
    raise Exception("Must be a string")
    #must be a string; answer was autopopulated

Exception: Must be a string

#### 3. What are the following errors?

In [7]:
arr = [1, 2, 3]
print(arr[3])

IndexError: list index out of range

the list index is out of range; must be [0:3] instead?

In [ ]:
fullName = "Taylor Swift"
print(full_name)

the variable name is wrong when trying to print the full name

In [ ]:
2 + '2' + 4

there is a mix of data with integers and text

## Functions

#### 1. Following the function design recipe, define a function that converts kilometers into miles. (Assume there are 1.6 kilometers in a mile.)

In [31]:
# Your code goes here
def convert_km_to_miles(km):
  """ Convert kilometers to miles. Assume there are 1.6 kilometers in a mile. """
  miles = km / 1.6
  return miles

In [32]:
convert_km_to_miles(8)

5.0

#### 2. What value is printed in the below code?

In [11]:
answer = 3

def answer_to_everything():
    '''Return the answer to life, the universe, and everything'''
    answer = 42
    return answer

answer_to_everything()

print(answer)
#the value that is printed in the below code is the variable "answer" which is the value 3

3


#### 3. Complete the body of the function. You can assume num is always >= 0.

In [25]:
def repeat(string, num):
    ''' Return string repeated num times.
    >>> repeat('yes', 4)
    'yesyesyesyes'
    >>> repeat('no', 0)
   #no will not run
    >>> repeat('yesnomaybe', 3)
    "yesnomaybeyesnomaybeyesnomaybe"
    '''
    output = 'num' >= 0
    return output

#### 4. Complete the function below. You can assume that number will always be a string with 10 digits and that the first 3 digits will always be the area code.

In [26]:
def is_416_number(number):
    '''
    Check whether the number has a 416 area code.
    >>> is_416_number('(416)-555-5555')
    True
    >>> is_416_number('514 416 5555')
    False
    >>> is_416_number('4165554160')
    True
    '''
    is_416_number.replace('()', ' ')
    is_416_number.replace('-', ' ')
    is_416_number.replace(' ', '')

    if number[0:3] == '416':
        return True
    else:
        return False
    # Hint: replace () and spaces with nothing
    #if else was autopopulated, but i would answer this question